nirtorch/interpreter.py

```diff
@@ -0,0 +1,210 @@
+
+
+def _default_map_affine(linear: nir.Affine) -> torch.nn.Linear:
+    module = torch.nn.Linear(linear.weight.shape[-1], linear.weight.shape[-2], bias=True)
+    module.weight.data = torch.from_numpy(linear.weight)
+    module.weight.bias = torch.from_numpy(linear.bias)
+    return module
+
+
+def _default_map_conv1d(conv: nir.Conv1d) -> torch.nn.Conv1d:
+    module = torch.nn.Conv1d(
+        conv.weight.shape[-3],
+        conv.weight.shape[-2],
+        conv.weight.shape[-1],
+        stride=conv.stride,
+        padding=conv.padding,
+        dilation=conv.dilation,
+        groups=conv.groups,
+        bias=True,
+    )
+    module.weight.data = torch.from_numpy(conv.weight)
+    module.bias.data = torch.from_numpy(conv.bias)
+    return module
+
+
+def _default_map_conv2d(conv: nir.Conv2d) -> torch.nn.Conv2d:
+    c = torch.nn.Conv2d(
+        conv.weight.shape[-4],
+        conv.weight.shape[-3],
+        conv.weight.shape[-2:],
+        stride=conv.stride,
+        padding=conv.padding,
+        dilation=conv.dilation,
+        groups=conv.groups,
+        bias=True,
+    )
+    c.weight.data = torch.from_numpy(conv.weight)
+    c.bias.data = torch.from_numpy(conv.bias)
+    return c
+
+
+def _default_map_linear(linear: nir.Linear) -> torch.nn.Linear:
+    module = torch.nn.Linear(linear.weight.shape[-1], linear.weight.shape[-2], bias=False)
+    module.weight.data = torch.from_numpy(linear.weight)
+    return module
+
+
+DEFAULT_MAP: NodeMapType = {
+    nir.Input: lambda input: torch.nn.Identity(),
+    nir.Output: lambda output: torch.nn.Identity(),
+    # Node mappings
+    nir.Affine: _default_map_affine,
+    nir.Conv1d: _default_map_conv1d,
+    nir.Conv2d: _default_map_conv2d,
+    nir.Linear: _default_map_linear,
+}
+
+
+def _map_nir_node_to_torch(
+    node: nir.NIRNode, node_map: NodeMapType
+) -> typing.Optional[torch.nn.Module]:
+    if type(node) in node_map:
+        return node_map[type(node)](node)
+    else:
+        raise ValueError(
+            f"Unknown node type {type(node)}, mapping does not exist in node_map"
+        )
+
+
+def _construct_module_dict_recursive(
+    nir_graph: nir.NIRGraph, node_map: NodeMapType
+) -> torch.nn.ModuleDict:
+    owning_module = torch.nn.ModuleDict()
+    for name, node in nir_graph.nodes.items():
+        # Recurse into subgraphs
+        if isinstance(node, nir.NIRGraph):
+            owning_module[name] = _construct_module_dict_recursive(node, node_map)
+        else:
+            mapped_module = _map_nir_node_to_torch(node, node_map=node_map)
+            if mapped_module is not None:
+                owning_module[name] = mapped_module
+    return owning_module
+
+
+def _find_result_names(
+    name: str, nir_graph: nir.NIRGraph, node_outputs: typing.Tuple[typing.Any]
+):
+    input_nodes = list(filter(lambda t: t[1] == name, nir_graph.edges))
+    return tuple(node_outputs[a] for a, _ in input_nodes)
+
+
+def _is_stateful(module: torch.nn.Module) -> bool:
```

Comment:
Given that you have an "or" at the end, does this really ensure you have a "state" argument in the function?  Just becaues `_construct_fx_graph` depends on this when you have a stateful function. 

Also, how would you enforce this in customizations to the frameworks, for example, if John Doe creates a brand new `AwesomeNeuron` type for `Norse`, can you raise an exception if the `forward` method lacks the "state" argument?